network/src/node.rs

```diff
@@ -119,106 +143,168 @@ impl<S: Storage + Send + Sync + 'static> Node<S> {
+                        //  to ensure some diversity in mempool state that is fetched.
+                        //  For now, this is acceptable because we propogate the mempool to
+                        //  all of our connected peers anyways.
+
+                        // The order of preference for the sync node is as follows:
+                        //   1. Iterate (in declared order) through the bootnodes:
+                        //      a. Check if this node is connected to the specified bootnode in the peer book.
+                        //      b. Select the specified bootnode as the sync node if this node is connected to it.
+                        //   2. If this node is not connected to any bootnode,
+                        //      then select the last seen peer as the sync node.
+
+                        // Step 1.
+                        let mut sync_node = None;
+                        for bootnode in bootnodes.iter() {
+                            if node_clone.peer_book.is_connected(*bootnode) {
+                                sync_node = Some(*bootnode);
+                                break;
+                            }
+                        }
 
-                    // make sure that the node doesn't remain in a sync state without peers
-                    if self_clone.state() == State::Syncing && self_clone.peer_book.number_of_connected_peers() == 0 {
-                        self_clone.set_state(State::Idle);
-                    }
+                        // Step 2.
+                        if sync_node.is_none() {
+                            // Select last seen node as block sync node.
+                            sync_node = node_clone.peer_book.last_seen();
+                        }
 
-                    // report node's current state
-                    trace!("Node state: {:?}", self_clone.state());
+                        sync_clone.update_memory_pool(sync_node).await;
+                    }
                 }
             });
-            self.register_task(state_mgmt_task);
-
-            if let Some(ref consensus) = self.consensus() {
-                let self_clone = self.clone();
-                let consensus = Arc::clone(consensus);
-                let transaction_sync_interval = consensus.transaction_sync_interval();
-                let tx_sync_task = task::spawn(async move {
-                    loop {
-                        sleep(transaction_sync_interval).await;
-
-                        if !consensus.is_syncing_blocks() {
-                            info!("Updating transactions");
-
-                            // select last seen node as block sync node
-                            let sync_node = self_clone.peer_book.last_seen();
-                            consensus.update_transactions(sync_node).await;
+            self.register_task(sync_mempool_task);
+
+            let node_clone = self.clone();
```

Comment:
the `Sync` object already contains a reference to `Node`, so it doesn't need to be additionally cloned here

Reply:
I think this would require changing the function signature from:
```rust
pub async fn start_services(&self)
```
to
```rust
pub async fn start_services(static self)
```

Reply:
I believe this would require `Node` to have a static lifetime in the method

Reply:
```bash
   --> network/src/node.rs:177:33
    |
177 |     pub async fn start_services(&self) {
    |                                 ^^^^^
    |                                 |
    |                                 this data with an anonymous lifetime `'_`...
    |                                 ...is captured here...
...
265 |             let sync_block_task = task::spawn(async move {
    |                                   ----------- ...and is required to live as long as `'static` here
```

Reply:
the node would be accessed via `sync_clone.node()` inside the task; that being said, it's not any blocker

Reply:
Ah good catch, I was using `sync`

Reply:
Fixed